File: 3ano/2Semestre/IART/baldes/node/Node.java
package node;

import java.util.Arrays;

import operation.Operation;

public class Node
{
	public int[] state;
	public int cost;
	public int level;
	public Node parent = null;
	public String operationName = null;

	public Node()
	{
		this.state = new int[2];
		this.cost = 0;
	}

	public Node(int state1, int state2, int cost)
	{
		this.state = new int[2];
		this.state[0] = state1;
		this.state[1] = state2;

		this.cost = cost;
	}

	public Node(Node node)
	{
		this.state = new int[2];
		this.state[0] = node.state[0];
		this.state[1] = node.state[1];

		this.cost = node.cost;
	}

	@Override
	public String toString() {
		return "Node {state=" + Arrays.toString(state) + ", cost=" + cost + "}";
	}
	
	
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + Arrays.hashCode(state);
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Node other = (Node) obj;
		if (!Arrays.equals(state, other.state))
			return false;
		return true;
	}
}

File: 3ano/2Semestre/IART/baldes/cpp/node.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

class Node
{
public:
	vector<int> state = {0, 0};
	int cost;
	int level;
	Node* parent = NULL;
	string operationName = "";

	Node()
	{
		this->cost = 0;
	}

	Node(int state1, int state2, int cost)
	{
		this->state[0] = state1;
		this->state[1] = state2;

		this->cost = cost;
	}

	Node(const Node& node)
	{
		this->state[0] = node.state[0];
		this->state[1] = node.state[1];

		this->cost = node.cost;
	}
	
	bool operator==(const Node* node) const
	{
		return (this->state == node->state);
	}

	bool operator==(const Node& node) const
	{
		return (this->state == node.state);
	}
};

inline ostream& operator<<(ostream& os, const Node& node)
{
	return (os << "Node {state=[" << node.state[0] << ", " << node.state[1] << "], " << "cost = " << node.cost);
}

struct hashNode
{
	int operator() (const Node* node) const
	{
		return static_cast<int>(std::hash<int>()(node->state[0]) + std::hash<int>()(node->state[1]));
	}

	bool operator() (const Node* node1, const Node* node2) const
	{
		return (node1->state == node2->state);
	}
};
